samples/nrf9160/modem_shell/src/location/location_service_nrf_cloud.c

```diff
@@ -0,0 +1,260 @@
+	if (ret == NULL) {
+		mosh_error("Failed to add pred int min to P-GPS request");
+		err = -ENOMEM;
+		goto cleanup;
+	}
+	ret = cJSON_AddNumberToObject(data_obj, NRF_CLOUD_JSON_PGPS_GPS_DAY,
+				      pgps_req->gps_day);
+	if (ret == NULL) {
+		mosh_error("Failed to add GPS day to P-GPS request");
+		err = -ENOMEM;
+		goto cleanup;
+	}
+	ret = cJSON_AddNumberToObject(data_obj, NRF_CLOUD_JSON_PGPS_GPS_TIME,
+				      pgps_req->gps_time_of_day);
+	if (ret == NULL) {
+		mosh_error("Failed to add GPS time to P-GPS request");
+		err = -ENOMEM;
+		goto cleanup;
+	}
+
+	/* Convert CJSON object to string and send to nRF Cloud */
+	msg_string = cJSON_PrintUnformatted(pgps_req_obj);
+	if (msg_string == NULL) {
+		mosh_error("Could not allocate memory for request message");
+		goto cleanup;
+	}
+
+	struct nrf_cloud_tx_data mqtt_msg = {
+		.data.ptr = msg_string,
+		.data.len = strlen(msg_string),
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE,
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE,
+	};
+
+	err = nrf_cloud_send(&mqtt_msg);
+	if (!err) {
+		mosh_print("P-GPS data requested");
+	}
+
+cleanup:
+	cJSON_Delete(pgps_req_obj);
+	cJSON_free(msg_string);
+
+	if (err) {
+		mosh_error("nRF Cloud P-GPS request failed, error: %d", err);
+	}
+}
+#endif /* CONFIG_NRF_CLOUD_PGPS */
+
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR) || defined(CONFIG_LOCATION_METHOD_WIFI)
+static void location_service_nrf_cloud_location_ready_cb(
+	const struct nrf_cloud_location_result *const result)
+{
+	if ((result != NULL) && (result->err == NRF_CLOUD_ERROR_NONE)) {
+		nrf_cloud_location.latitude = result->lat;
+		nrf_cloud_location.longitude = result->lon;
+		nrf_cloud_location.accuracy = (double)result->unc;
+
+		k_sem_give(&location_ready);
+	} else {
+		if (result) {
+			mosh_error(
+				"Unable to determine location from cloud response, error: %d",
+				result->err);
+		}
+		/* Reset the semaphore to unblock location_service_nrf_cloud_pos_get()
+		 * and make it return an error.
+		 */
+		k_sem_reset(&location_ready);
+	}
+}
+
+static int location_service_nrf_cloud_pos_get(
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+	const struct lte_lc_cells_info *cell_data,
+#else
+	void *cell_data,
+#endif
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+	const struct wifi_scan_info *scanning_results,
+#else
+	void *scanning_results,
+#endif
+	struct location_data *const location)
+{
+	int err;
+	nrf_cloud_location_response_t callback = NULL;
+
+	k_sem_reset(&location_ready);
+
+	if (location) {
+		callback = location_service_nrf_cloud_location_ready_cb;
+	}
+
+	mosh_print("Sending positioning request to cloud via MQTT");
+	err = nrf_cloud_location_request(
+		cell_data, scanning_results, (callback != NULL), callback);
+	if (err == -EACCES) {
+		mosh_error("Cloud connection is not established");
+		return err;
+	} else if (err) {
+		mosh_error("Failed to request positioning data, error: %d", err);
+		return err;
+	}
+
+	if (callback) {
+		if (k_sem_take(&location_ready, K_MSEC(30000)) == -EAGAIN) {
```

Comment:
We are having  a timeouts in a location command.

Reply:
Yep, I guess I need to calculate the remaining timeout in `location_shell.c` and pass it to `location_service_cellular_ext_handle()` (and WiFi)

Reply:
Yes, timeout handling for ext architecture of location lib would be good to have as a sample.

Reply:
Right now I've only added a comment above this `k_sem_take()` as discussed offline.
Using --cellular_timeout (or --wifi_timeout) would require making it global in location_shell.c and passing here through few function calls. We could even count how long was between `location_request()` and `LOCATION_EVT_CELLULAR_EXT_REQUEST`.
And even that wouldn't be enough as we would need to take `CONFIG_LOCATION_REQUEST_DEFAULT_CELLULAR_TIMEOUT` into account by reading the cellular timeout after a call to `location_config_defaults_set()`. All this is possible but not very nice and clean considering what is it for.

Looking at it from another perspective, we can use here an independent cloud transaction related timeout which conceptually should be fine. And that's now this 30s timeout.